src/main/java/org/commcare/formplayer/services/CaseSearchHelper.java

```diff
@@ -0,0 +1,51 @@
+package org.commcare.formplayer.services;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.commcare.formplayer.screens.FormplayerQueryScreen;
+import org.commcare.modern.util.Pair;
+import org.javarosa.core.model.instance.ExternalDataInstance;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.cache.CacheManager;
+import org.springframework.cache.annotation.CacheConfig;
+import org.springframework.cache.annotation.CacheEvict;
+import org.springframework.cache.annotation.Cacheable;
+import org.springframework.http.HttpEntity;
+import org.springframework.http.HttpHeaders;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.ResponseEntity;
+import org.springframework.lang.Nullable;
+import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.web.client.RestTemplate;
+
+import java.io.ByteArrayInputStream;
+import java.io.UnsupportedEncodingException;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+
+@CacheConfig(cacheNames = {"case_search"})
+public class CaseSearchHelper {
+
+    @Autowired
+    CacheManager cacheManager;
+
+    @Autowired
+    QueryRequester queryRequester;
+
+    @Cacheable(unless="#result == null", key="#username + #uri")
+    public ExternalDataInstance getSearchDataInstance(String username, FormplayerQueryScreen screen, String uri, HttpHeaders headers) {
+        String responseString = queryRequester.makeQueryRequest(uri, headers);
+        if (responseString != null) {
+            Pair<ExternalDataInstance, String> dataInstanceWithError = screen.processResponse(
+                    new ByteArrayInputStream(responseString.getBytes(StandardCharsets.UTF_8)));
+            return dataInstanceWithError.first;
+        }
+        return null;
+    }
+
+    @CacheEvict(allEntries = true)
```

Comment:
we should be able to do this with the cache configuration `spring.cache.caffeine.spec=expireAfterWrite=30m`. We could apply that globally to all caches or we could configure the caches separately (https://www.baeldung.com/spring-multiple-cache-managers)

Reply:
@snopoke Lookds like configuring a cache manager in Java code will override the `.properties` config and will enable the cache by default without having the ability to disable the cache using `.properties` file. Therefore I am going to lean towards applying a global config for all caches to `expireAfterWrite` after 5 mins.

Reply:
I think that's fine for the current use cases. The ability to disable it was only intended for initial rollout so it would also be fine to have them permanently configured. Or if we really want we can take an approach like this: https://medium.com/@d.lopez.j/configuring-multiple-ttl-caches-in-spring-boot-dinamically-75f4aa6809f3